### ISSUE TITLE
throws on first select when a stream exists on that transaction

### ISSUE THREAD

**Issue by Sil3ntStorm** [jtv/libpqxx#401]
Trying to query some data from the database (which in this case won't exist, which is expected), however I am getting an exception that the transaction was aborted.

```
auto tx = make_unique<work>(db_con);
stream_to strm(*tx, "foo_bar", vector<string>{"id", "time", "val"});
for (const auto& value : data) {
    if (auto result = tx->exec_params("SELECT id FROM bar WHERE name = $1", someValue); result.size() == 0) { // throws
        // do other stuff
    }
   // more code
}
```

That is pretty much what is being executed, and it is the first activity on the database. If the stream is not there, it works as expected.
However I need to query values from the database while doing the loop, so I do need both of them on the same connection.

**Comment by jtv** [comment]
Can you determine at what point exactly the error occurs, and what exactly the error message is?

**Comment by jtv** [comment]
Just noticed the change to this ticket's title...  You can't execute queries on a transaction while it has a stream open, but I would expect a different error message for that.  Can you clarify?

**Comment by Sil3ntStorm** [comment]
Well the exec_params on the transaction simply throws an pqxx::sql_error which contains the following message: "ERROR:  current transaction is aborted, commands ignored until end of transaction block"

**Comment by jtv** [comment]
That helps, thanks.  I agree that's not very helpful.  I intend to do two things about it:
 1. Document clearly that you can't execute normal queries while you have a stream open.  There are several such restrictions at the protocol and libpq levels; in libpqxx I model them as "transaction focus" classes.  A transaction can have only one "focus" at a time, and `stream_from` is one of these classes.
 2. Internally represent SQL statements as transaction focus objects as well, so that you get a clearer error message.  When you try to create a "focus" object on a transaction that already has an active focus, libpqxx throws an exception with a message like "attempt to open X while Y still active."

**Comment by Sil3ntStorm** [comment]
Talking about documentation though (I know we all hate having to actually do those), I noticed that there is quite a bit more information in the actual header files with regards to comments, than is actually displayed by the IDE or even in the doxygen generated html.
Maybe we could get those changed to something that is actually processed and displayed (preferably directly in the IDE) or at least in the doxygen output.
For example none of the exec functions tell you what it does (or some of the pitfalls might be), even though there is some information in the header file.

**Comment by jtv** [comment]
Sometimes I documented a function once but did not repeat the same documentation (or nearly the same) for all its variants.  I'll have to see if I can do something with groupings.

**Comment by jtv** [comment]
@Sil3ntStorm I've added documentation for a bunch of functions, and in the cases where multiple functions needed the same documentation, I've created groups.  I also rejigged the guard against executing SQL commands while some other transaction focus is already active.

Could you go over the problems and see whether they have improved?  I have a feeling that an exception may be getting suppressed somewhere.

**Comment by Sil3ntStorm** [comment]
I will see that I find some time to look into the latest commit within the next 20 hours.

**Comment by Sil3ntStorm** [comment]
```
try {
    auto tx = make_unique<work>(db_con);
    stream_to strm(*tx, "foo_bar", vector<string>{"id", "time", "val"});
    for (const auto& value : data) {
        if (auto result = tx->exec_prepared("someStatement", someValue); result.size() == 0) { // throws
            // do other stuff
        }
        // more code
    }
} catch (const sql_error& e) {
    cerr << e.sqlstate() << "::" << e.what() << '\n';
}
```
yields the following error message: `25P02::Failure during 'someStatement': ERROR:  current transaction is aborted, commands ignored until end of transaction block`

using commit c052a7804bcb500ab0aa6353021b9d099f73fe3c

**Comment by jtv** [comment]
I just reproduced the problem...  Turns out I _did_ check for open streams and such before executing a statement, but _not_ before executing a prepared statement or a parameterised statement.  Which internally are entirely different.

@Sil3ntStorm I hope this fixes it for you.  It's still an error to execute a statement while a `stream_to` (or a `stream_from` or a `pipeline`) is active, but now you get an error message that actually says so.
